cadence/server/api.go

```diff
@@ -0,0 +1,252 @@
+	}
+}
+
+// /api/request/id
+// Posts a request submission for a specific song ID.
+func RequestID() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		clog.Info("Request", fmt.Sprintf("Request-by-ID by client %s.", r.Header.Get("X-Forwarded-For")))
+
+		type Request struct {
+			ID string `json:"ID"`
+		}
+		var request Request
+
+		decoder := json.NewDecoder(r.Body)
+		err := decoder.Decode(&request)
+		if err != nil {
+			clog.Error("RequestID", "Unable to decode request.", err)
+			w.WriteHeader(http.StatusBadRequest) // 400 Bad Request
+			return
+		}
+		reqID, err := strconv.Atoi(request.ID)
+		if err != nil {
+			clog.Error("RequestID", "Unable to convert request ID to an integer.", err)
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		path, err := getPathById(reqID)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		_, err = liquidsoapRequest(path)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+
+		w.WriteHeader(http.StatusAccepted) // 202 Accepted
+	}
+}
+
+// /api/request/bestmatch
+// Posts a request submission for the top result of a search.
+func RequestBestMatch() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		clog.Debug("Search", fmt.Sprintf("Decoding http-request data from client %s.", r.Header.Get("X-Forwarded-For")))
+
+		type RequestBestMatch struct {
+			Query string `json:"Search"`
+		}
+		var rbm RequestBestMatch
+
+		decoder := json.NewDecoder(r.Body)
+		err := decoder.Decode(&rbm)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest) // 400 Bad Request
+			return
+		}
+
+		queryResults, err := searchByQuery(rbm.Query)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		path, err := getPathById(queryResults[0].ID)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		_, err = liquidsoapRequest(path)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+
+		w.WriteHeader(http.StatusAccepted) // 202 Accepted
+	}
+}
+
+// /api/nowplaying/metadata
+// Gets text metadata (excludes album art and path) of the currently playing song.
+func NowPlayingMetadata() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		queryResults, err := searchByTitleArtist(nowTitle, nowArtist)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		if len(queryResults) < 1 {
+			w.WriteHeader(http.StatusNotFound) // 404
+			return
+		}
+		song := queryResults[0]
+
+		// Return data to client
+		type SongData struct {
+			ID     int
+			Artist string
+			Title  string
+			Album  string
+			Genre  string
+			Year   int
+		}
+		result := SongData{ID: song.ID, Artist: song.Artist, Title: song.Title, Album: song.Album, Genre: song.Genre, Year: song.Year}
+		jsonMarshal, _ := json.Marshal(result)
```

Comment:
Finally getting rid of the whole `parseMusic` pseudo-JSONP that neither the 4.0 frontend or CadenceBot handled the way JSONP is supposed to be used.

... Are there angels singing, or is it my imagination?